### DISCUSSION TITLE
oauth 2 refresh in app router

### DISCUSSION THREAD

**Discussion by alighafoorzade** [vercel/next.js#68170] [App Router]
Let's imagine we have server side component in which we get data that is protected and user needs login to access it. something like below:
`
"use server"
const SSRComponent = async()=> {
    const data = await getData();
}
`
and i as i said, the data is protected and user needs to login to access it and it's possible to access token expires during navigation between pages and you should refresh the access token and set refresh and access token. something like the code below:
`
const getData = async()=>{
    const {access, refresh} = await refreshToken();
    cookies().set('refresh', refresh);
    cookies().set('access', access);
}
`
but the problem is that you can't set cookies in the the ssr component. so what is the proper way and performant way to solve this issue becaus calling api call in middleware blocks user from seeing the page and based on server response time, this can affect first contentfull paint and largest contentfull paint.

**Comment by icyJoseph** [comment]
`use server` doesn't mark stuff as Server Components, rather it is a way back for `use client` modules, to invoke server side code at runtime. 

Components are Server Components by default, there's no directive for that, unless marked with a `use client` directive. 

You can try to use the `import 'server-only'` package, which causes a crash when code that should be only for the server, is somehow included in a client module. https://www.npmjs.com/package/server-only

When your user does login, you do this in a router handler, or server action, and then likely set cookies or something, that's not during SSR. And to refresh the user cookie, you could use middleware, and even do it within server actions or route handlers. It should be rather cheap to check if a user session has expired though, only those who need to refresh their session, would wait a bit longer.

**Reply by alighafoorzade** [reply]
@icyJoseph thanks for your response, could you provide a sample code for refreshing in middleware, middleware for pages or for endpoints? btw i should call external api to get refresh and access token, so this the reason that it takes time and middleware could took up to 300ms to respond.
